source/postprocess/visualization.cc

```diff
@@ -127,6 +127,82 @@ namespace aspect
           }
       };
 
+
+      template <int dim>
+      class BaseAdjointVariablePostprocessor: public DataPostprocessor< dim >, public SimulatorAccess<dim>
+      {
+        public:
+          virtual
+          void
+          evaluate_vector_field(const DataPostprocessorInputs::Vector<dim> &input_data,
+                                std::vector<Vector<double> > &computed_quantities) const
+          {
+            const double velocity_scaling_factor =
+              this->convert_output_to_years() ? year_in_seconds : 1.0;
+            const unsigned int n_q_points = input_data.solution_values.size();
+            for (unsigned int q=0; q<n_q_points; ++q)
+              for (unsigned int i=0; i<computed_quantities[q].size(); ++i)
+                {
+                  // scale velocities and fluid velocities by year_in_seconds if needed
+                  if (this->introspection().component_masks.velocities[i] ||
+                      (this->include_melt_transport()
+                       && this->introspection().variable("fluid velocity").component_mask[i]))
+                    computed_quantities[q][i]=input_data.solution_values[q][i] * velocity_scaling_factor;
+                  else
+                    computed_quantities[q][i]=input_data.solution_values[q][i];
+                }
+          }
+
+          virtual std::vector<std::string> get_names () const
+          {
+
+            std::vector<std::string> solution_names (dim, "adjoint_velocity");
+
+            if (this->include_melt_transport())
+              {
+                solution_names.push_back ("p_f");
+                solution_names.push_back ("p_c_bar");
+                for (unsigned int i=0; i<dim; ++i)
+                  solution_names.push_back ("u_f");
+              }
+            solution_names.push_back ("adjoint_p");
+            solution_names.push_back ("adjoint_T");
+//             for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+//                 solution_names.push_back (this->introspection().name_for_compositional_index(c));
+
+            return solution_names;
+          }
+
+
+          virtual
+          std::vector<DataComponentInterpretation::DataComponentInterpretation>
+          get_data_component_interpretation () const
+          {
+            std::vector<DataComponentInterpretation::DataComponentInterpretation>
+            interpretation (dim,
+                            DataComponentInterpretation::component_is_part_of_vector);
+            if (this->include_melt_transport())
```

Comment:
Are you suggesting that the lines 162:167 should also stay in? Following Rene's suggestion, I now check in the parameter file whether the calculation includes melt while also using the adjoint solver and if so to throw an exception. I'm not sure if it would break things if you include melt, but it might?

Reply:
I think a general principle is that if you extend code like you do, and there is an easy way to support a complex scenario in one place, then you do that even though maybe you don't (and assert) in another place. Here, you can just import the data component interpretation flags from the primal solution like I showed for the names. I would do that, and that will work with or without melt transport. That doesn't mean that you have to support melt transport in all other places you change, and you should add asserts in those places that melt transport isn't enabled. But here, it's easy and it is one place less that someone would have to change later on when adding melt to adjoints.

Now, I don't think anyone will ever come around and combine melt and adjoints into anything that is actually useful. So the question is academic for this particular case, but I thought I'd explain my thinking because you're going to run into the same kind of question in other places and contexts as well.